### ISSUE TITLE
Problem with showing IGV pileup for CRAM served from Nebula

### ISSUE THREAD

**Issue by tonydisera** [iobio/gene.iobio#1127]
Asking Ralph R. for a URL to a CRAM to troubleshoot this problem...

**Comment by zorgster** [comment]
I have a Nebula CRAM file... and it hangs when I press the Read Pileup link (I've not logged in for a while, so this is new to me).. 

I have subsetted my CRAM file to gene STAT3.. and subsetted my VCF to the same region... this much smaller file also hangs when I press 'Read Pileup'.
I have converted the file to BAM and it still hangs.

[In order to read the CRAM file I had to download GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz.. use samtools seq_cache_populate.pl script to create a cache..  and then set environment variables that samtools uses to use the reference cache..  No other GRCh38 worked... but that wouldn't affect the BAM file... ]

Both the BAM and the CRAM load into https://igv.org/ and appear the same.

**Comment by tonydisera** [comment]
Thank you @zorgster for verifying that another Nebula user has this same problem. It would be a great help if you could send me your subsetted BAM for gene STAT3 to me for troubleshooting. Please email me directly to coordinate. 

tony.disera@gmail.com

Best regards,
Tony

**Comment by zorgster** [comment]
The dependency for igv in vue-pileup is "igv": "^2.2.13".  In the modules folder, I see IGV 2.4.0.  Could this have been a bug in 2.4.0?

In the history for IGV - a few changes spotted - https://github.com/igvteam/igv.js/releases:
3.0.5 - a bug fix in 3.0.5 in which some actions could cause the browser to freeze when loading large regions..   
2.13.8 - BAM index optimisations
2.13.3 - The URIs in CRAM files did not work.. https://github.com/igvteam/igv.js/pull/1548

**Comment by tonydisera** [comment]
Wow, Oliver @zorgster, that is impressive troubleshooting! I bet you nailed it. I am away from my desk, but will give this a try when I am back at my computer. The IGV pileup is working  for some bams/crams web served, but it could be failing under specific conditions triggered by the newer version of igv.js. Many thanks!

**Comment by zorgster** [comment]
I have loaded the igvjs@3.1.1 project, and edited their example "cram-vcf.html" file to load hg38 + the STAT3.vcf and .cram ... everything loaded ok.

I've cloned Vue-pileup... edited the file to load up the cram file...  Could this be relevant for the CRAM files?

Error: CRAM version 67 not supported
    at new e (igv.esm.js:18358:1)
    at t.eval (igv.esm.js:18385:1)
    at w (igv.esm.js:18385:1)
    at Generator.eval [as _invoke] (igv.esm.js:18385:1)
    at t.<computed> [as next] (igv.esm.js:18385:1)
    at n (igv.esm.js:18378:1)
    at eval (igv.esm.js:18378:1)

**Comment by zorgster** [comment]
Vue-pileup seems to be failing to detect the BAI as an index file.  Is it reading the bytes in little endian?  Are there two problems?  one for CRAM (version 67?) and the other for the BAM index .. BAI?

Using this code (in bash) - it confirms the 'magic number' is 21578050 for BAI which matches the number in the IGV.esm.js file in the loadIndex function.   Although it must be fine because I can use it in other IGV products.

```
magic=$(xxd -l 4 -p STAT3.bam.bai | awk '{print "0x" substr($0,7,2) substr($0,5,2) substr($0,3,2) substr($0,1,2)}')
magic_int=$((magic))
echo $magic_int

21578050
```

**Comment by tonydisera** [comment]
Many thanks, @zorgster, for your excellent troubleshooting! I'm going to bring @anderspitman into the conversation. He wrote the Vue.js component that wraps igv.js.  

@anderspitman, we have a recent problem with viewing IGV pileup for certain bams and crams. This was noticed by some of our Nebula users.  I may need your help figuring out why the pileup component is failing for these bams/crams. The good news is that we have this incredible user @zorgster, who has kindly spent time narrowing in on the problem. He has provided a subsetted bam that we can use to troubleshoot as well. Can you think of any changes to the pileup component that would have resulted in a 'CRAM version 67 not supported' error message or '..... is not a bai file'?

**Comment by anderspitman** [comment]
It looks like vue-pileup hasn't changed in 5 years, so if this is a new problem I suspect it has to do with the data files. Maybe the file formats have evolved and we're now out of date?

It's interesting that it fails even after you update igv.js though. One thing to note is that when using the pileup dialog it's going to use the igv.js embedded in vue-pileup, but when opening in a separate tab it's going to navigate you to igv.iobio.io, which has it's own (old) version. I honestly can't remember why I did it this way but it seems like a terrible design now.

**Comment by tonydisera** [comment]
Interesting. Well, I won't troubleshoot that way! 

This is strange that read pileup works on the demo data in gene.iobio. But seems to freeze with Oliver's provided file.

**Comment by zorgster** [comment]
I spoke too soon - I have restarted from a fresh clone and now I can load these into Vue-pileup - I may have altered the IGV version unwittingly.

![image](https://github.com/user-attachments/assets/328cbfd4-968b-42f6-9861-fdc22ffd6aea)

**Comment by tonydisera** [comment]
Thanks @zorgster, that is good info! I'm communicating with @anderspitman on Slack, so I will pass this along.

**Comment by zorgster** [comment]
I used the following for the Proband track...

```
        {
          name: 'Proband',
          alignmentURL: "http://localhost:8080/STAT3.cram",
          alignmentIndexURL: "http://localhost:8080/STAT3.cram.crai",
          variantURL: "http://localhost:8080/STAT3.vcf.gz",
        },

```

And for Pileup:

```
    <pileup
      heading="Read pileup BALLY SNP 17:42322301"
      referenceURL="https://s3.amazonaws.com/igv.broadinstitute.org/genomes/seq/hg38/hg38.fa"
      indexURL="https://s3.amazonaws.com/igv.broadinstitute.org/genomes/seq/hg38/hg38.fa.fai"
      locus="chr17:42322290-42322330"
      :tracks='tracks'
    />

```

**Comment by tonydisera** [comment]
> I spoke too soon - I have restarted from a fresh clone and now I can load these into Vue-pileup - I may have altered the IGV version unwittingly.
> 
> ![image](https://private-user-images.githubusercontent.com/4734627/394874605-328cbfd4-968b-42f6-9861-fdc22ffd6aea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM5NDYwMDcsIm5iZiI6MTczMzk0NTcwNywicGF0aCI6Ii80NzM0NjI3LzM5NDg3NDYwNS0zMjhjYmZkNC05NjhiLTQyZjYtOTg2MS1mZGMyMmZmZDZhZWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTFUMTkzNTA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGM5MzMxZjU3MDhhNmJlMDA1NmVhNGRmZTQ0MDhkNDliMDc4NTgyYWQ3OTRlNGMxNWZlODYwMTllODg4NmUyMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.oRxNwnUaCb0T-z0G-V11psFHmAGrEbLZ4pN-UwS3yCI)

Hi @zorgster, what version of igv.js in your local build of vue-pileup is used? I'm seeing the same 'Magic number' problem you are w/ my local vue-pileup build. Also, I am curious what version of node you used to build vue-pileup. I had to use v16, but this could be a local environment issue.

**Comment by zorgster** [comment]
It wasn't clear which nvm I needed... I ran `nvm use 13` before every npm command from within VSCode because it kept slipping back to 23..   I ran it with 13.   Other versions of IGV I loaded were 2.6.0, 2.6.6 and 3.1.1... 

Along the way I also spotted that there is a version change from CRAM v3.0 to v3.1... (around 2022). 
 https://academic.oup.com/bioinformatics/article/38/6/1497/6499262 ...

But I'm back thinking the bam and cram loading in gene.iobio may have a similar cause..  In gene.iobio, they are uploaded to the FL Proxy?  and then used from that location?  Mine worked when they were on http on the same location... if that helps...?

**Comment by anderspitman** [comment]
@tonydisera noticed that even though the latest version of vue-pileup on npm is 0.11.0, the repo is only 0.10.0. Apparently I never push the latest changes. I did a diff on the two versions:

```
Only in 0.10: lib
diff -r 0.10/package.json 0.11/package.json
3c3
<   "version": "0.10.0",
---
>   "version": "0.11.0",
11c11
<     "igv": "^2.2.13",
---
>     "igv": "^2.4.1",
diff -r 0.10/src/App.vue 0.11/src/App.vue
27a28
>           variantIndexURL: "https://s3.amazonaws.com/iobio/samples/vcf/platinum-exome.vcf.gz.tbi",
33a35
>           variantIndexURL: "https://s3.amazonaws.com/iobio/samples/vcf/platinum-exome.vcf.gz.tbi",
diff -r 0.10/src/components/Pileup.vue 0.11/src/components/Pileup.vue
31c31
< import igv from '../../lib/igv.esm'
---
> import igv from 'igv'
162a163
>       indexURL: tracks[0].variantIndexURL,
166c167
<   const url = 'https://s3.amazonaws.com/static.iobio.io/dev/igv.iobio.io/index.html?config=' + JSON.stringify(igvConfig)
---
>   const url = 'https://igv.iobio.io?config=' + JSON.stringify(igvConfig)
```

**Comment by anderspitman** [comment]
Are we feeling confident that the latest version of igv.js is working? If so, and to avoid build issues in the future, maybe we should just make a new vanilla JS web component and wrap the latest version. The wrapper is very minimal anyway and Vue plays nicely with web components.

**Comment by zorgster** [comment]
I cloned the current igv.js repo, edited the cram-vcf.html and it loaded it with the VCF, BAM and CRAM all at once...

**Comment by tonydisera** [comment]
> Are we feeling confident that the latest version of igv.js is working? If so, and to avoid build issues in the future, maybe we should just make a new vanilla JS web component and wrap the latest version. The wrapper is very minimal anyway and Vue plays nicely with web components.

Yes! This is a great suggestion, @anderspitman . We may need the vue-pileup component for tighter integration in other apps, like our new rnasplice.iobio.iobio, but for our use case in gene.iobio, this would be the way to go.

**Comment by tonydisera** [comment]
> I cloned the current igv.js repo, edited the cram-vcf.html and it loaded it with the VCF, BAM and CRAM all at once... (via npx `http-server` ) (on node 13.14.0) ...with the BAM, CRAM and VCF files in the same folder as the examples/cram-vcf.html file.

Thank you @zorgster. I build with node v13.14.0 for gene.iobio. I can reproduce the bug in vue-pileup v 0.10 with the STAT3.bam file you provided. It works with all other files I have tried. I verified that the proxy is work on other files, so we can eliminate that variable.

**Comment by zorgster** [comment]
Can I just check... The full CRAM file from Nebula is 50 GB for 30X... does this get uploaded to the proxy?  I suspect that this wouldn't be possible in the brief time it takes to load the website.  How does this work?

**Comment by anderspitman** [comment]
Yeah bioinformatics tools usually don't work at all with servers that don't support range requests. The files tend to just be too big.

**Comment by AlistairNWard** [comment]
No, there shouldn't be. 150bp reads are pretty standard, but can get a lot longer than that

**Comment by zorgster** [comment]
Hi...
I have been adding console.log points into the code to isolate where it stops.

I added one before and after extractQuery in IGV.js and the 'after extractQuery was not hit.  So i set a few inside. 

This image is where my browser appears frozen - but the hits on "eq tokens.length " is still increasing... 

![image](https://github.com/user-attachments/assets/41db005f-e588-4f06-8c19-34071f77ec27)

**Comment by zorgster** [comment]
The issue here is the href where sex0 has no '=' and the code is trying to split the variables by '='...

because tokens.length == 1, i is never incremented in the for loop...

```
	    for (i = i1 + 1; i < i2;) {
	      j = uri.indexOf("&", i);
	      if (j < 0) j = i2;
	      s = uri.substring(i, j);
	      tokens = s.split("=", 2);

        console.log(" i1 " + i1 + " - i2 " + i2 + " - i " + i + " - j " + j);
        console.log("eq s = " + s);
	      if (tokens.length === 2) {
	        key = tokens[0];
	        value = decodeURIComponent(tokens[1]);

	        if ('file' === key) {
	          // IGV desktop style file parameter
	          files = value.split(',');
	        } else if ('index' === key) {
	          // IGV desktop style index parameter
	          indexURLs = value.split(',');
	        } else {
	          config[key] = value;
	        }

	        i = j + 1; 
	      }
	    }

```

http://localhost:4026/?gene=STAT3&genes=STAT3&species=Human&build=GRCh38&affectedSibs=&unaffectedSibs=&rel0=proband&sex0&vcf0=https%3A%2F%2Flf-proxy.iobio.io%2Fpyvh-vd5u%2FBALLYBROOM.vcf.gz&tbi0=https%3A%2F%2Flf-proxy.iobio.io%2Fpyvh-vd5u%2FBALLYBROOM.vcf.gz.tbi&bam0=https%3A%2F%2Flf-proxy.iobio.io%2Fb04f-mv5c%2FBALLYBROOM.cram&bai0=https%3A%2F%2Flf-proxy.iobio.io%2Fb04f-mv5c%2FBALLYBROOM.cram.crai&sample0=BALLYBROOM&affectedStatus0=affected

**Comment by zorgster** [comment]
Also i see:

```
Access to XMLHttpRequest at 'https://lf-proxy.iobio.io/3p2m-60em/STAT3.cram?someRandomSeed=0.31t1siraqkk' from origin 'http://localhost:4026' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

igv.js:25634 
GET https://lf-proxy.iobio.io/3p2m-60em/STAT3.cram?someRandomSeed=0.31t1siraqkk net::ERR_FAILED 206 (Partial Content)
```

**Comment by tonydisera** [comment]
Hi @zorgster. Thank you so much for your excellent troubleshooting! I apologize for not getting back to you sooner. I only worked a few days last week. 
#
The browser freezing issue is - if Sex is not selected by the user in gene.iobio - in the href, sex0 has no '=' and the code is trying to split the variables by '='...

You have saved me so much time by figuring out the root cause of the problem. I was wrongly assuming that a new version of igv.js would get us closer to solving the problem. 

As for the CORS problem, my guess is that lf-proxy.iobio.io allows for the official production domains of gene.iobio, both for the public version and the Nebula version, so we should be okay.

I will change the errant code in gene.iobio so that the URL argument for sex is only present when it is known. (I'm assuming the igv.js problem goes away when sex is provided, but I will verify this.)

Thanks again!!

Tony

**Comment by zorgster** [comment]
Hi @tonydisera 

I can confirm that the browser does not hang if I select a gender when loading the files...

However, as the problem was addressed after 3.0.0 in igv.js, updating igv.js would also solve the problem, without having to alter the code around the Sex0 parameter.

I used the production server at gene.iobio.io with my cram file.. it also gave the CORS error in the Pileup modal window, but not in the main gene.iobio app.  Which could be caused by the encoding of the URL + query parameter (randomSeed)..?  Something like this?

https://stackoverflow.com/questions/78083319/cors-error-only-when-using-a-url-query-string-parameter

Oliver

**Comment by tonydisera** [comment]
Hi @zorgster,

I apologize for the radio silence. I took time off for the holidays and then had to travel for a funeral. But now I am back and will give this issue my full attention. 

I did try specifying the sex in gene.iobio as you pointed out in your post. And I see that it does work. (Excellent!) Now I am a bit stuck on updating to igv.js to verify the fix. For some reason, I cannot get the wrapper app for vue-pileup to work with the cram you provided. I must be missing something. So I will spend a few more hours trying to validate that the fix works with an upgrade to igv.js. 

I am grateful and impressed that you were able to troubleshoot this. I will keep you posted on this issue. 

Thank you,
Tony

**Comment by tonydisera** [comment]
Hi @zorgster,

I wanted to reach out to you and let you know that I have created a new version of the Vue wrapper around igv.js. I have verified that the latest igv.js (3.2) (in the Vue.js component used in the gene.iobio app) indeed handles the edge-case URL parsing problem (missing = delimiter to tokenize parameter substring).

I have been absent from work for the last few months dealing with a health complication. I am now recovered and back to work. I want to express my gratitude to you. You saved me many hours of troubleshooting and dug deep into the underlying igv.js code. You pinpointed the function and lines of code in the older igv.js version that threw an error caused by incorrect parsing of the window/browser URL string. And you cleverly tracked down the gene.iobio use-case that resulted in the unexpected URL 'sex' parameter. Which explained the older igv.js version's missing logic to handle the edge case of null or missing elements in the array of strings that were tokenized assuming that the delimiter '=' would be present and the string token after the '=' delimiter would be present. You did that. And you found the fixed function in the latest version of igv.js 3.2. Thank you.

After your heavy lifting, I have spent many hours trying to build and run igv.js 3.2. In igv 3.0, the code utilizes ES2020 syntax and requires a browser that supports ES6. For a Vue.js app that depends on the ECMA module, it must transpile the module with proper babel configuration. Now add in the complicated interplay of babel, webpack, and vue configuration, generated from either vue-cli or vite with different approaches and level of support for vue.js 2 vs vue.js 3. Why does a module with modern ES2020 javascript syntax require transpilation? I am not entirely sure of the reason or even it is necessary. I will try to find out.  someday.  Enough complaining. 

TLDR - Thank you for your expertise and generosity. There were extenuating circumstances that delayed a new release of gene with igv.js 3.2. It is forthcoming (another week or so?). The public gene.iobio is separate from the Nebula, so there may be an additional delay for Nebula customers clicking on the Nebula site link.
